src/headers/frame_header.rs

```diff
@@ -0,0 +1,494 @@
+    #[default(0)]
+    #[condition(have_crop)]
+    height: u32,
+
+    /* "normal_frame" denotes the condition !all_default
+    && (frame_type == kRegularFrame || frame_type == kSkipProgressive) */
+    #[default(BlendingInfo::default())]
+    #[condition(frame_type == FrameType::RegularFrame || frame_type == FrameType::SkipProgressive)]
+    #[nonserialized(num_extra_channels : nonserialized.num_extra_channels,
+        have_crop : have_crop, x0: x0, y0: y0, width: width, height: height,
+        img_width: nonserialized.img_width, img_height: nonserialized.img_height)]
+    blending_info: BlendingInfo,
+
+    #[size_coder(explicit(nonserialized.num_extra_channels))]
+    #[default(vec![BlendingInfo::default(); nonserialized.num_extra_channels as usize])]
+    #[nonserialized(num_extra_channels : nonserialized.num_extra_channels,
+        have_crop : have_crop, x0: x0, y0: y0, width: width, height: height,
+        img_width: nonserialized.img_width, img_height: nonserialized.img_height)]
+    ec_blending_info: Vec<BlendingInfo>,
+
+    #[coder(u2S(0, 1, Bits(8), Bits(32)))]
+    #[default(0)]
+    #[condition((frame_type == FrameType::RegularFrame ||
+        frame_type == FrameType::SkipProgressive) && nonserialized.have_animation)]
+    duration: u32,
+
+    #[coder(Bits(32))]
+    #[default(0)]
+    #[condition((frame_type == FrameType::RegularFrame ||
+        frame_type == FrameType::SkipProgressive) && nonserialized.have_timecode)]
+    timecode: u32,
+
+    #[default(frame_type == FrameType::RegularFrame)]
+    #[condition(frame_type == FrameType::RegularFrame || frame_type == FrameType::SkipProgressive)]
+    pub is_last: bool,
+
+    #[coder(Bits(2))]
+    #[default(0)]
+    #[condition(frame_type != FrameType::LFFrame && !is_last)]
+    save_as_reference: u32,
+
+    /* The default value for save_before_ct is d_sbct = !(full_frame && (frame_type ==
+    kRegularFrame || frame_type == kSkipProgressive) && blending_info.mode ==
+    kReplace && (duration == 0 || save_as_reference != 0) && !is_last). */
+    #[default(!(!have_crop || (x0 == 0 && y0 == 0 &&
+        width + x0 >= nonserialized.img_width && height + y0 >= nonserialized.img_height))
+        && (frame_type == FrameType::RegularFrame || frame_type == FrameType::SkipProgressive) &&
+        blending_info.mode == BlendingMode::Replace && (duration == 0 || save_as_reference != 0) && !is_last)]
+    #[condition(frame_type != FrameType::LFFrame)]
+    save_before_ct: bool,
+
+    name: String,
+
+    #[default(RestorationFilter::default())]
+    #[nonserialized(encoding : encoding)]
+    restoration_filter: RestorationFilter,
+
+    #[default(Extensions::default())]
+    extensions: Extensions,
+}
+
+#[cfg(test)]
+mod test_frame_header {
+    use super::*;
+    use crate::{
+        bit_reader::BitReader,
+        bmff::JxlCodestream,
+        headers::{FileHeaders, JxlHeader},
+    };
+
+    macro_rules! test_frame_header {
```

Comment:
Is there a specific reason why this is a macro and not a function?

Reply:
Yeah, it's better to use a function here, I originally wrote it in a way which wouldn't be possible with a function, but forgot about it.